content/Refactoring - Sebastian Herrera.md

```diff
@@ -0,0 +1,144 @@
+
+**Conclusion: Pay attention to code smells. They might be whispering (or even shouting!) asking for a better solution.**
+
+## 4) ROLE OF TESTING
+By definition, refactoring is the process of changing the code structure, without altering its behavior.
+- How do we know that we have not altered its behavior then?
+  - Yes, you’ve guessed. We need tests.
+- And what if the code we want to refactor doesn’t have enough tests?
+  - Then you should first write the tests, and then refactor.
+
+You may be asking other things like whether it's better to have more unit tests or integration tests, or other things like if it's necessary to have near 100% code coverage on tests.
+
+The short answer is, there is no specific technique or condition to be met in order to be sure that your code works exactly as expected. There is a quote from Dijkstra that says *"Program testing can be used to show the presence of bugs, but never to show their absence."*
+
+So test the way you prefer, and try to cover as much as possible without adding unnecessary duplicated tests. Keep in mind that this will allow you to apply refactoring with much more confidence.
+
+**Conclusion: Having tests is a precondition for refactoring, unless you want to take the risk of breaking things.**
+
+## 5) MECHANICS OF A REFACTORING
+After this long introduction, we are finally reaching the part where I present a couple of examples. Before doing that, I would like to mention that the refactoring as described by Martin Fowler in his book, they have a format, and they have a mechanics.
+
+Format of a refactoring:
+- They have a name
+  - Is important to building a vocabulary of refactoring.
+- They have a motivation
+  - Describes why the refactoring should be done and describes circumstances in which it shouldn’t be done.
+- They have a mechanic
+  - Step-by-step description of how to carry out the refactoring.
+
+Mechanics of a refactoring:
+- Each step in the refactoring is as small as possible.
+- Test after every step. 
+- In practice, if you want to take larger steps, you can, but if the tests fail, then you should go backwards and take smaller steps.
+
+In the book they are presented within 6 different categories which may not be worth it to explain them in this article. Here you can find a [catalog of refactoring](https://refactoring.com/catalog/).
+
+**Conclusion: There is no conclusion here.**
+
+## 6) SOME EXAMPLES
+If you read up to here, first thank you, and second, now you have to finish it.
+
+I will present just 2 simple refactoring as examples. I don’t think the purpose of this article should be for you to learn many of them or learning some complex ones. I believe that is something you can learn by yourself in somewhere else, but you will do it only if you feel enough motivation to do it.
```

Comment:
```suggestion
I will present just 2 simple refactorings as examples. I don’t think the purpose of this article should be for you to learn many of them or learning some complex ones. I believe that is something you can learn by yourself somewhere else, but you will do it only if you feel enough motivation to do it.
```

Reply:
Nice, I thought that `refactoring` was the plurar too, however it seems that `refactorings` is valid.